stencila/pyla/servers.py

```diff
@@ -0,0 +1,289 @@
+    MethodNotFound = -32601
+
+    """Invalid method parameter(s)."""
+    InvalidParams = -32602
+
+    """Internal JSON-RPC error."""
+    InternalError = -32603
+
+    """Generic server error."""
+    ServerError = -32000
+
+    """Capability error."""
+    CapabilityError = -32005
+
+
+class JsonRpcError(BaseException):
+    """
+    A JSON-RPC error that may be part of a response
+
+    Python implementation of Executa's
+    [JsonRpcError](https://github.com/stencila/executa/blob/v1.0.0/src/base/JsonRpcError.ts).
+    """
+
+    """
+    A number that indicates the error type that occurred.
+    This MUST be an integer.
+    """
+    code: JsonRpcErrorCode
+
+    """
+    A string providing a short description of the error.
+    The message SHOULD be limited to a concise single sentence.
+    """
+    message: str
+
+    """
+    A primitive or structured value that contains additional information about the error.
+    This may be omitted.
+    The value of this member is defined by the server (e.g. detailed error information,
+    nested errors etc.).
+    """
+    data: typing.Any
+
+    def __init__(self, code: JsonRpcErrorCode, message: str, data: typing.Any = None):
+        super().__init__(message)
+        self.code = code
+        self.message = message
+        self.data = data
+
+
+class StreamServer:
+    """
+    A server that communicates using length-prefixed JSON-RPC messages over streams or sockets.
+
+    Python implementation of Executa's
+    [StreamServer](https://github.com/stencila/executa/blob/v1.0.0/src/stdio/StreamServer.ts#L10)
+    """
+
+    interpreter: Interpreter
+    input_stream: StreamType
+    output_stream: StreamType
+
+    def __init__(self, interpreter: Interpreter, input_stream: StreamType, output_stream: StreamType) -> None:
+        self.interpreter = interpreter
+        self.input_stream = input_stream
+        self.output_stream = output_stream
+
+    def read_message(self) -> typing.Iterable[str]:
+        """Read a length-prefixed message from the input stream then repeat."""
+        while True:
+            yield message_read(self.input_stream)
+
+    def write_message(self, message: str) -> None:
+        """Write a length-prefixed message to the output stream."""
+        message_write(self.output_stream, message)
+
+    def receive_message(self, message: str) -> str:
+        """
+        Receive a JSON-RPC request and send back a JSON-RPC response.
+
+        The response may have a JSON-RPC `error` if the request was bad.
+
+        Python implementation of Executa's
+        [Server.receive](https://github.com/stencila/executa/blob/v1.0.1/src/base/Server.ts#L61).
+        """
+        request_id = None
+        result = None
+        error = None
+
+        try:
+            try:
+                request = json.loads(message)
+            except BaseException as exc:
+                raise JsonRpcError(JsonRpcErrorCode.ParseError, 'Parse error: %s' % repr(exc))
```

Comment:
This should use new style string formatting:
`raise JsonRpcError(JsonRpcErrorCode.ParseError, 'Parse error: {}'.format(exc))`

I _think_ the `repr` call is redundant in this case.

Reply:
OK, thanks